papers/vyas_ramasubramani/94_vyas_ramasubramani.rst

```diff
@@ -329,10 +329,13 @@ To ensure that the state point associated with the job id can be recovered, a JS
 
 This storage mechanism enables :math:`O(1)` access to the data associated with a particular state point through its hash, but more importantly, parsing of the JSON state point files in the workspace allows :math:`O(N)` indexing of the data space.
 This indexing is performed by traversing the data space and parsing the state point files directly, and other files may also be parsed along the way.
-This mode of operation is well-suited to the high performance filesystems common to high performance computing.
-For better performance, the resulting indexes can be manually saved; in general, ``signac`` automatically caches generated indexes for this purpose.
+In general, ``signac`` automatically caches generated indexes within a single session where possible, but for better performance the indexes can also be manually saved.
 These indexes then allow efficient selection and searching of the data space, and MongoDB-style queries can be used for complex selections.
 
+This distributed mode of operation is well-suited to the high performance filesystems common to high performance computing.
+The implicit partitioning of the data on a per-job basis ensures that any operations that do not require explicit indexing will be fast and nearly scale-invariant with respect to the data space size.
+At present, ``signac`` does not support explicit horizontal partitioning of the data space, but because a ``signac`` project can always be linked with a new workspace, manually partitioning the data space so as to emulate sharding is a relatively simple task that simply requires the user to determine how the data should be divided between shards.
```

Comment:
Since the state point represents the primary key and if we interpret the document as one row within the table, then the signac data layout is even for one workspace essentially horizontally partitioned. However, I think that partitioning the data manually is less relevant here, than the fact that the data is inherently distributed and may be accessed in parallel by potentially thousands of clients in parallel. This is why it suits typical HPC workflows very well.

How about this:
> The explicit horizontal partitioning and distributed storage of data per job is well suited to typical HPC operations, which are executed for multiple jobs in parallel.
> Since data is accessed distributively there is no inherent bottleneck posed by funneling all data read and write operations through one or more server applications.

Reply:
I agree that for our purposes the inherent distribution of the data is the important point, especially for HPC. However, the particular question this was meant to address was relating to sharding, which typically encompasses horizontal partitioning across filesystems. While one benefit is to reduce load, which is automatically addressed by signac as is, another is potentially to reduce the total index size, or perhaps to make better use of available space. In that context, it would be helpful to indicate how the data space can be actually partitioned. I like your phrasing, but I think it would be good to incorporate a sentence discussing that aspect as well, which is what my last sentence was meant to address. Do you think we can combine the two versions?